src/utils/scheduler.hpp

```diff
@@ -76,55 +75,44 @@ class Scheduler {
         // waiting that function f will not log before it.
         // Check for pause also.
         auto lk = std::unique_lock{mutex_};
-        auto now = std::chrono::system_clock::now();
         next_execution += pause;
+        auto now = std::chrono::system_clock::now();
         if (next_execution > now) {
-          condition_variable_.wait_until(lk, next_execution, [&] { return !is_working_.load(); });
+          condition_variable_.wait_until(lk, next_execution, [&] { return token.stop_requested(); });
         } else {
           next_execution = find_next_execution(now);  // Compensate for time drift when using a start time
         }
 
-        pause_cv_.wait(lk, [&] { return !is_paused_.load(); });
+        pause_cv_.wait(lk, [&] { return !is_paused_.load(std::memory_order_acquire) || token.stop_requested(); });
+
+        if (token.stop_requested()) break;
 
-        if (!is_working_) break;
         f();
       }
     });
   }
 
   void Resume() {
-    is_paused_.store(false);
+    is_paused_.store(false, std::memory_order_release);
     pause_cv_.notify_one();
   }
 
-  void Pause() { is_paused_.store(true); }
+  void Pause() { is_paused_.store(true, std::memory_order_release); }
 
-  /**
-   * @brief Stops the thread execution. This is a blocking call and may take as
-   * much time as one call to the function given previously to Run takes.
-   * @throw std::system_error
-   */
   void Stop() {
-    is_paused_.store(false);
-    is_working_.store(false);
-    pause_cv_.notify_one();
-    condition_variable_.notify_one();
-    if (thread_.joinable()) thread_.join();
+    if (thread_.request_stop()) {
+      is_paused_.store(false, std::memory_order_release);
+      pause_cv_.notify_one();
+      condition_variable_.notify_one();
+      if (thread_.joinable()) thread_.join();
+    }
```

Comment:
If two threads call Stop, only one will actually execute the stop and that is correct.
But the other one will fall-through. I would expect this to be blocking.
Or add a mechanism that blocks until joinable is false or just document well.

Reply:
I think falling-through is expected honestly. If you try to stop stopped think, why would you block? I documented, i think that's enough.

Reply:
sure